lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
+            Exactly one of 'tag' or 'branch' must be specified for a git url. 'commit'
+            may be provided in addition to either 'tag' or branch' to ensure that the
+            commit matches the given string.
+            Given:
+            commit={0}
+            tag={1}
+            branch={2}
+            """).format(commit, tag, branch))
+        if commit is not None:
+            return cls.Commit(commit)
+        if tag is not None:
+            return cls.Tag(tag)
+        assert branch is not None, (commit, tag, branch, version_name)
+        return cls.Branch(branch)
+
+    def refspec(self):
+        # type: () -> str
+        """Return a string which unambiguously identifies this ref to git."""
+        if self.ref_type == 'Commit':
+            return self._ref
+        if self.ref_type == 'Tag':
+            return 'refs/tags/{0}'.format(self._ref)
+        assert self.ref_type == 'Branch', self
+        return 'refs/heads/{0}'.format(self._ref)
+
+    def fetch_spec(self):
+        # type: () -> str
+        """Return an argument to update a local ref from the remote, if applicable."""
+        if self.ref_type == 'Commit':
+            # Commits cannot be updated and do not require any special syntax to fetch.
+            return self.refspec()
+        # Otherwise, we use the syntax to update the local ref from a remote ref of
+        # the same name, fetching updates if available/applicable.
+        assert self.ref_type in ['Tag', 'Branch'], self
+        maybe_mutating_prefix = '+' if self.is_mutable() else ''
+        return '{0}{1}:{1}'.format(maybe_mutating_prefix, self.refspec())
+
+    def is_mutable(self):
+        # type: () -> bool
+        """Whether this type of git ref should check for updates to that ref."""
+        if self.ref_type == 'Branch':
+            return True
+        assert self.ref_type in ['Tag', 'Commit'], self
+        return False
+
+    def _repo_info(self):
+        # type: () -> str
+        if self.ref_type == 'Commit':
+            return 'at commit {0}'.format(self._ref)
+        if self.ref_type == 'Tag':
+            return 'at tag {0}'.format(self._ref)
+        assert self.ref_type == 'Branch', self
+        return 'on branch {0}'.format(self._ref)
+
+    def repo_info_for_reference(self):
+        # type: () -> str
+        if self.should_validate_matches_hash:
+            return '{} (at commit {})'.format(
+                self._repo_info(), self.should_validate_matches_hash)
+        return self._repo_info()
+
+
+class GitCommit(GitRef):
+    """A git commit."""
+
+    ref_type = 'Commit'        # type: ClassVar[str]
+
+    # Matches any 7-40 character hexadecimal string.
+    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
+
+    @property
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        """There is no need to verify that a git commit hash matches itself."""
+        return None
```

Comment:
Won't just return (without None) still return None? is there an old Python version that doesn't support that?

Reply:
Yes, but the return value is going to be read by another function, so I wanted to make it clear it wasn't going to be thrown away. You're right that that's already somewhat indicated by the type annotation `-> Optional[str]`.